### PR TITLE
Detection matrix input helper function

### PR DESCRIPTION
**Pull request by Purg**
Added the discussed helper function to assist in forming the specific input requested by the `GenerateDetectorProposalSaliency.generate()` interface method.

### CODE REVIEW

**Review by brianhhu — COMMENTED** [review]
LGTM, added a few minor comments.

File: xaitk_saliency/utils/detection.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        shape `[nDets x 4]`. The format of each row-vector is not important
+        but generally expected to be `[left, top, right, bottom]` pixel
+        coordinates. This matrix will must be of a type that is float-castable.
+    :param classification_mat: Matrix of classification results from the
+        detector or detection classifier. This should have the shape of
+        `[nDets x nClasses]`. This matrix must be of a type that is
+        float-castable.
+    :param objectness: Optional vector of objectness scores for input
+        detections. This is optional as not all detection models output this
+        aspect. When provided, this should be a vector of ints/floats of size
```
[the remark this answers is not in the document]
Comment (Purg) [reply xaitk_saliency/utils/detection.py:44]: Mostly because numpy handles the type casting minimization for us. Also it doesn't matter how the values are represented as long as they are within the desired range. You do raise a hidden point that I could add the [0,1] inclusive range expectation.

File: tests/utils/test_detection.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        # all be 1's.
+        assert set(combined_mat[:, 4]) == {1}
+
+    def test_explicit_objectness(self) -> None:
+        """
+        Test that input objectness vector is represented in the output.
+        """
+        test_bbox_mat = np.random.random_integers(0, 255, 16 * 4).reshape(16, 4)
+        test_class_mat = np.random.rand(16, 8)
+        test_obj_v = np.tile([1, 2, 3, 4], 4)
```
[the remark this answers is not in the document]
Comment (Purg) [reply tests/utils/test_detection.py:27]: Valid. I can adjust this to use decimals instead of integers.

File: examples/DRISE.ipynb [hunk 3]
```diff
@@ -385,10 +388,11 @@
     "    boxes = dets['boxes'].cpu().numpy()\n",
     "    scores = dets['scores'][:, valid_idxs].cpu().numpy()\n",
     "    labels = dets['labels'].cpu().numpy()\n",
-    "    \n",
-    "    # Add dummy objectness scores for D-RISE\n",
-    "    objectness = np.ones((boxes.shape[0], 1))\n",
-    "    output = np.concatenate((boxes, objectness, scores), axis=1)\n",
+    "    # no \"objectness\" scores from this model.\n",
+    "\n",
+    "    # This helper function makes assists in making sure the output is in the format\n",
```
[the remark this answers is not in the document]
Comment (Purg) [reply examples/DRISE.ipynb:393]: Done.

File: xaitk_saliency/utils/detection.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Purg) [reply xaitk_saliency/utils/detection.py:44]: I added a note about our expectation that "objectness" is valied in the [0,1] range in the `GenerateDetectorProposalSaliency` interface method doc-string.

File: xaitk_saliency/utils/detection.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+    In exceptions about shape mismatches, index 0 refers to the `bbox_mat`
+    input, index 1 refers to the ``objectness` vector, and index 2 refers to
+    the `classification_mat`.
+
+    :param bbox_mat: Matrix of bounding boxes. This matrix should have the
+        shape `[nDets x 4]`. The format of each row-vector is not important
+        but generally expected to be `[left, top, right, bottom]` pixel
+        coordinates. This matrix will must be of a type that is float-castable.
+    :param classification_mat: Matrix of classification results from the
```
Comment (brianhhu) [inline xaitk_saliency/utils/detection.py:38]: Not sure if this should be clarified here, but `classification_mat` is essentially the confidences/logits from the classification, so these will also be in the range [0,1].

```suggestion
    :param classification_mat: Matrix of classification scores from the
```

**Review by brianhhu — APPROVED** [review]
LGTM, I added a few comments about providing additional detail on classification_mat. Feel free to ignore if you don't think are necessary.

**Review by schencej — APPROVED** [review]
LGTM. Just found a couple typos.

File: xaitk_saliency/interfaces/gen_detector_prop_sal.py [hunk 5]
```diff
@@ -35,18 +35,24 @@ def generate(
...
-        scores, we recommend using a value of 1.0 in that column.
+        our minimum requirements of a bounding-box, per-class scores.
+        Objectness scores are required in our input format, but not
+        necessarily from detection black-box methods as there is a sensible
+        default value for this.
+        See the :py:func:`~.utils.detection.format_detection` helper function
+        for assistance in forming our input format, which includes this
+        optional default fill-in.
+        We expect objectness is a confidence score valued in the inclusive
+        ``[0,1]`` range.
```
[the remark this answers is not in the document]
Comment (Purg) [reply xaitk_saliency/interfaces/gen_detector_prop_sal.py:46]: I'll add it. here since we're in those weeds.

File: tests/utils/test_detection.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+    def test_input_nonimpact(self) -> None:
+        """
+        Test that the invocation of the function does not change the input.
+        """
+        def gen_fresh_bbox() -> np.ndarray:
+            return np.tile([1, 2, 3, 4], 16).reshape(16, 4)
+
+        def gen_fresh_clss() -> np.ndarray:
+            return np.tile([5, 6, 7, 8, 9, 0], 16).reshape(16, 6)
```
[the remark this answers is not in the document]
Comment (Purg) [reply tests/utils/test_detection.py:55]: fixed

File: xaitk_saliency/utils/detection.py (hunk 4 above)
Comment (Purg) [reply xaitk_saliency/utils/detection.py:38]: fixed

File: xaitk_saliency/utils/detection.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    If an `objectness` score vector is not provided, we assume a vector of 1's.
+
+    The output of this function is a matrix that is of shape
+    `[nDets x (4+1+nClasses)]`. This is the result of horizontally stacking the
+    input in bbox, objectness and classification order. The output matrix
+    data-type will follow numpy's rules about safe-casting given the
+    combination of input matrix types.
+
+    In exceptions about shape mismatches, index 0 refers to the `bbox_mat`
+    input, index 1 refers to the ``objectness` vector, and index 2 refers to
```
[the remark this answers is not in the document]
Comment (Purg) [reply xaitk_saliency/utils/detection.py:31]: I will argue that this is not a typo as this get's rendered as RST content, and the intent here is to mark this word as *literals*, which uses double-backticks, where as the other content here that uses a single backtick are rendered as italicized (except for domain syntax, like `:py:meth:`, that is explicitly supposed to use single-backticks). If this is ever rendered as markdown, markdown rendered double-backticked content as literals as well in addition to the markdown-usual single-backticks.

I will grant you that I could correct the other occurrences here of the ticks so that we are consistent.

[the remark this answers is not in the document]
Comment (schencej) [reply xaitk_saliency/utils/detection.py:31]: So...?
```suggestion
    input, index 1 refers to the ``objectness`` vector, and index 2 refers to
```

[the remark this answers is not in the document]
Comment (Purg) [reply xaitk_saliency/utils/detection.py:31]: Yup, I will have an updated fixing this in a moment.

File: xaitk_saliency/utils/detection.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    the `classification_mat`.
+
+    :param bbox_mat: Matrix of bounding boxes. This matrix should have the
+        shape `[nDets x 4]`. The format of each row-vector is not important
+        but generally expected to be `[left, top, right, bottom]` pixel
+        coordinates. This matrix will must be of a type that is float-castable.
```
[the remark this answers is not in the document]
Comment (Purg) [reply xaitk_saliency/utils/detection.py:37]: Fixed

**Comment by brianhhu** [comment]
LGTM, thanks all! Feel free to merge on green.
